tiledb/sm/query/dimension_label/dimension_label_range_query.h

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file dimension_label_range_query.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Classes for creating queries to read ranges.
+ */
+
+#ifndef TILEDB_DIMENSION_LABEL_RANGE_QUERY_H
+#define TILEDB_DIMENSION_LABEL_RANGE_QUERY_H
+
+#include "tiledb/common/common.h"
+#include "tiledb/sm/query/query.h"
+#include "tiledb/sm/storage_manager/storage_manager_declaration.h"
+#include "tiledb/type/range/range.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+
+class DimensionLabel;
+class IndexData;
+class QueryBuffer;
+class Subarray;
+
+/**
+ * Return a Status_DimensionQueryRangeError error class Status with a given
+ * message.
+ *
+ * Note: This intentionally returns the error as TileDB::DimensionLabelQuery.
+ **/
+inline Status Status_DimensionLabelRangeQueryError(const std::string& msg) {
+  return {"[TileDB::DimensionLabelQuery] Error", msg};
+}
+
+class DimensionLabelRangeQuery {
+ public:
+  /** Destructor. */
+  virtual ~DimensionLabelRangeQuery() = default;
+
+  /**
+   * Retrieves the computed ranges from the query.
+   *
+   * @returns [is_point_range, range_data, count]
+   *   * is_point_range: If ``true`` the returned data is stored as point
+   * ranges, otherwise it stored as [start, end] range pairs.
+   *   * range_data: Pointer to the start of the range data.
+   *   * count: Total number of points stored in the range data.
+   */
+  virtual tuple<bool, const void*, uint64_t> computed_ranges() const = 0;
+
+  /** Returns ``true`` if the query status is completed. */
+  virtual bool completed() const = 0;
+
+  /** Processes the dimension label query. */
+  virtual void process() = 0;
+};
+
+class OrderedRangeQuery : public DimensionLabelRangeQuery {
+ public:
+  /** Default constructor is not C.41 compliant. */
+  OrderedRangeQuery() = delete;
```

Comment:
Is this needed because that object is used in a container that needs default constructors?

Reply:
The containers that use it are all containers of pointers, not directly on this object. I could allow default construction of this class, but if I do, I would like to allow the default constructor in the `DimensionLabelDataQuery` subclasses as well to be consistent.

The default constructed classes would not really be useable, but all the private members do have a reasonable default value (nullptr).  Should I allow the default constructor in all or none of them?

Reply:
Oh wow, I totally misread this 😅 We could we change the comment to say "Delete default constructor for C.41 compliance", but not necessary. When I read this, I though that this was defining a default constructor hence was not yet C.41 compliant.

Reply:
I'll reword that in the future.